contracts/contracts/GrantRoundManager.sol

```diff
@@ -0,0 +1,171 @@
+
+  constructor(
+    GrantRegistry _registry,
+    ISwapRouter _router,
+    IERC20 _donationToken
+  ) {
+    require(_registry.grantCount() >= 0, "GrantRoundManager: Invalid registry");
+    require(address(_router).isContract(), "GrantRoundManager: Invalid router"); // Router interface doesn't have a state variable to check
+    require(_donationToken.totalSupply() > 0, "GrantRoundManager: Invalid token");
+
+    registry = _registry;
+    router = _router;
+    donationToken = _donationToken;
+  }
+
+  /**
+   * @notice Creates a new GrantRound
+   * @param _owner Grant round owner that has permission to update the metadata pointer
+   * @param _payoutAdmin Grant round administrator that has permission to payout the matching pool
+   * @param _registry Address that contains the grant metadata
+   * @param _startTime Unix timestamp of the start of the round
+   * @param _endTime Unix timestamp of the end of the round
+   * @param _metaPtr URL pointing to the grant round metadata
+   * @param _minContribution Miniumum donation amount that can be made using the given donation token
+   */
+  function createGrantRound(
+    address _owner,
+    address _payoutAdmin,
+    GrantRegistry _registry,
+    uint256 _startTime,
+    uint256 _endTime,
+    string memory _metaPtr,
+    uint256 _minContribution
+  ) external {
+    GrantRound _grantRound = new GrantRound(
+      _owner,
+      _payoutAdmin,
+      _registry,
+      donationToken,
+      _startTime,
+      _endTime,
+      _metaPtr,
+      _minContribution
+    );
+
+    emit GrantRoundCreated(address(_grantRound));
+  }
+
+  /**
+   * @notice Swap and donate to a grant
+   * @param  _donation Donation being made to a grant
+   */
+  function swapAndDonate(Donation calldata _donation) external payable {
+    // --- Validation ---
+    // Rounds must be specified
+    GrantRound[] calldata _rounds = _donation.rounds;
+    require(_rounds.length > 0, "GrantRoundManager: Must specify at least one round");
+
+    // Only allow value to be sent if the input token is WETH
+    IERC20 _tokenIn = _donation.tokenIn;
+    require(
+      (msg.value == 0 && address(_tokenIn) != WETH) || (msg.value > 0 && address(_tokenIn) == WETH),
+      "GrantRoundManager: Invalid token-value pairing"
+    );
+
+    // Wnsure grant recieving donation exists in registry
+    uint96 _grantId = _donation.grantId;
+    require(_grantId < registry.grantCount(), "GrantRoundManager: Grant does not exist in registry");
+
+    // Iterate through each GrantRound to verify:
+    //   - The round has the same donationToken as the GrantRoundManager
+    //   - The round is active
+    for (uint256 i = 0; i < _rounds.length; i++) {
+      require(
+        donationToken == _rounds[i].donationToken(),
+        "GrantRoundManager: GrantRound's donation token does not match GrantRoundManager's donation token"
+      );
+
+      require(
+        block.timestamp >= _rounds[i].startTime() && block.timestamp < _rounds[i].endTime(),
```

Comment:
Does it make sense to add a helper method to the `GrantRound` like `isActive()` or `status()` or something similar, rather than redefining what it means to be "active" here?

Reply:
I agree but shouldn't the GrantRound contract be reverting here instead? What other contracts depend on this status state variable?

Reply:
> shouldn't the GrantRound contract be reverting here instead

Mmmm I'm not necessarily sure I agree with this. Why would we want the Round to revert? I was thinking along the lines of a public view method.

> What other contracts depend on this status state variable?

Good question. I'm not sure any right now, but it just seems useful to have the Round define canonically when it is active, as opposed to anything that integrates with the Round having to repeatedly define that logic itself.

Reply:
We don't actually call into the `GrantRound` anymore, so agreed with @apbendi that we shouldn't expect it to revert

> Does it make sense to add a helper method to the `GrantRound` like `isActive()` or `status()` or something similar, rather than redefining what it means to be "active" here?

Yea I like that idea, will do!

Reply:
Done in 79cbe2a